In [2]:
import numpy as np
import tensorflow as tf
import os # Directory navigation
import random
import wave # Read wav file

c:\Users\I27645\.conda\envs\python\lib\site-packages\scipy\__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 2.0.2
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"

A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.0.2 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "c:\Users\I27645\.conda\envs\python\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "c:\Users\I27645\.conda\envs\python\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "c:\Users\I27

AttributeError: _ARRAY_API not found


A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.0.2 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "c:\Users\I27645\.conda\envs\python\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "c:\Users\I27645\.conda\envs\python\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "c:\Users\I27645\.conda\envs\python\lib\site-packages\ipykernel_launcher.py", line 16, in <module>
    app.launch_new_instance()
  File "c:\Users\I27645\.conda\envs\python\lib\site-packages\traitlets\config\application.py", line 846, in launch_instance
    app.start()
  File

AttributeError: _ARRAY_API not found


A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.0.2 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "c:\Users\I27645\.conda\envs\python\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "c:\Users\I27645\.conda\envs\python\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "c:\Users\I27645\.conda\envs\python\lib\site-packages\ipykernel_launcher.py", line 16, in <module>
    app.launch_new_instance()
  File "c:\Users\I27645\.conda\envs\python\lib\site-packages\traitlets\config\application.py", line 846, in launch_instance
    app.start()
  File

AttributeError: _ARRAY_API not found


A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.0.2 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "c:\Users\I27645\.conda\envs\python\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "c:\Users\I27645\.conda\envs\python\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "c:\Users\I27645\.conda\envs\python\lib\site-packages\ipykernel_launcher.py", line 16, in <module>
    app.launch_new_instance()
  File "c:\Users\I27645\.conda\envs\python\lib\site-packages\traitlets\config\application.py", line 846, in launch_instance
    app.start()
  File

AttributeError: _ARRAY_API not found


A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.0.2 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "c:\Users\I27645\.conda\envs\python\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "c:\Users\I27645\.conda\envs\python\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "c:\Users\I27645\.conda\envs\python\lib\site-packages\ipykernel_launcher.py", line 16, in <module>
    app.launch_new_instance()
  File "c:\Users\I27645\.conda\envs\python\lib\site-packages\traitlets\config\application.py", line 846, in launch_instance
    app.start()
  File

AttributeError: _ARRAY_API not found

# 1. Dataset Preparation:
Use the Google Speech Commands Dataset and add noise variants. Optionally, for bonus points, extend the dataset to support additional languages by integrating public datasets (e.g., French, Spanish).


In [3]:
# Set the path to your main dataset directory
dataset_path = './archive (1)'

# Get a list of available command/folder names
commands = np.array(tf.io.gfile.listdir(str(dataset_path)))
print(commands)

['down' 'go' 'left' 'no' 'off' 'on' 'README.md' 'right' 'stop'
 'testing_list.txt' 'up' 'validation_list.txt' 'yes']


We want to keep: down, go, left, right, stop, up
Others to consider: yes/no, numbers, on/off

Note: LICENSE, README.md, testing_list.txt, validation_list.txt are documents and _background_noise is additional data.
Note: At this time, I went in and deleted all unneeded directories

In [4]:
commands = ["down", "go", "left", "right", "stop", "up", "yes", "no", "on", "off"] # commands to keep

# get file counts for each command
# Loop through each category and count the number of audio files
for command in commands:
    command_path = os.path.join(dataset_path, command)
    # Get the list of all .wav files in the command folder
    audio_files = [file for file in os.listdir(command_path)]
    print(f"Number of audio files in '{command}': {len(audio_files)}")

Number of audio files in 'down': 2359
Number of audio files in 'go': 2372
Number of audio files in 'left': 2353
Number of audio files in 'right': 2367
Number of audio files in 'stop': 2380
Number of audio files in 'up': 2375
Number of audio files in 'yes': 2377
Number of audio files in 'no': 2375
Number of audio files in 'on': 2367
Number of audio files in 'off': 2357


Each command has ~2375 examples.

From README:
The data was captured in a
variety of formats, for example Ogg Vorbis encoding for the web app, and then
converted to a 16-bit little-endian PCM-encoded WAVE file at a 16000 sample
rate. The audio was then trimmed to a one second length to align most
utterances, using the
extract_loudest_section
tool. The audio files were then screened for silence or incorrect words, and
arranged into folders by label.

Divided into directories this way, you can easily load the data using `keras.utils.audio_dataset_from_directory`. 

The audio clips are 1 second or less at 16kHz. The `output_sequence_length=16000` pads the short ones to exactly 1 second (and would trim longer ones) so that they can be easily batched.

## Add noise
Instead of adding noise during  preprocessing, we will add it via a tensorflow pipeline.

In [5]:
def read_wav_file(filepath):
    # Reads a WAV file and returns its sample rate and audio tensor
    with wave.open(filepath, 'rb') as wav: # Open the wav file
        sample_rate = wav.getframerate() # Get sample rate
        frames = wav.readframes(wav.getnframes()) # Get frames
        audio = np.frombuffer(frames, dtype=np.int16) # Get audio
    return sample_rate, tf.convert_to_tensor(audio, dtype=tf.float32) 

def sample_noise(noise_audio, target_length=1600): # double check inputs
    # Samples a random segment of noise to match the target audio length (1600 samples).
    noise_length = len(noise_audio) # Get length of audio
    if target_length > noise_length: # Check if audio is long enough for sample length
        raise ValueError("Noise file is too short for sampling.")
    
    start_idx = random.randint(0, noise_length - target_length) # Generate random int to start the audio sample from
    return noise_audio[start_idx:start_idx + target_length] # Create and return the audio sample

def add_noise_to_audio(audio, noise_audio, noise_level=0.1): # How to decide noise_level?
    # Adds noise to the audio at the specified noise level
    scaled_noise = noise_audio * noise_level
    return audio + scaled_noise

def augment_with_noise(audio, noise_audio, noise_level=0.1):
    # Augments a single audio example with noise
    noise_sample = sample_noise(noise_audio, len(audio))
    return add_noise_to_audio(audio, noise_sample, noise_level)

def create_dataset(audio_dir, noise_audio_path, batch_size=32, noise_level=0.1):
    #Creates a dataset with noise augmentation for 1-second, 1600-sample audio files.

    # Load noise audio once
    _, noise_audio = read_wav_file(noise_audio_path)

    # Load dataset
    train_ds, val_ds = tf.keras.utils.audio_dataset_from_directory(
        audio_dir,
        batch_size=batch_size,
        validation_split=0.2, # 80-20 split
        seed=0, # Set seed
        output_sequence_length=1600,  # Enforce uniform shape
        labels="inferred",
        subset="both"
    )

    # Add noise augmentation
    def add_noise(audio, label):
        audio = tf.numpy_function(
            lambda x: augment_with_noise(x, noise_audio, noise_level),
            [audio],
            tf.float32
        )
        return audio, label

    # Apply noise augmentation for both sets
    train_ds = train_ds.map(add_noise)
    val_ds = val_ds.map(add_noise)
    
    return train_ds, val_ds


## Create the pipeline with batchsize 32 and noise_level=0.2(play w/ this?)

In [ ]:
noise_file_path = "background/_background_noise_/exercise_bike.wav" # path background noise file

train, test = create_dataset(
    audio_dir=dataset_path, # Commands audio directory
    noise_audio_path=noise_file_path, # Background noise
    batch_size=32,
    noise_level=0.2
)


Found 23682 files belonging to 10 classes.
Using 18946 files for training.
Using 4736 files for validation.


# Split Data

In [ ]:
train.m

<bound method DatasetV2.enumerate of <_MapDataset element_spec=(TensorSpec(shape=<unknown>, dtype=tf.float32, name=None), TensorSpec(shape=(None,), dtype=tf.int32, name=None))>>